bpf/kmesh/workload/include/backend.h

```diff
@@ -30,52 +30,74 @@ static inline backend_value *map_lookup_backend(const backend_key *key)
 
+    value_tuple.ipv4.daddr = ctx->user_ip4;
+    value_tuple.ipv4.dport = ctx->user_port;
+
+    ret = bpf_map_update_elem(&map_of_dst_info, &sk, &value_tuple, BPF_NOEXIST);
+    if (ret) {
+        BPF_LOG(ERR, BACKEND, "record metadata origin address and port failed, ret is %d\n", ret);
+        return ret;
+    }
+    target_addr.ipv4 = addr;
+    target_addr.port = port;
+    SET_CTX_ADDRESS(ctx, target_addr);
+    kmesh_workload_tail_call(ctx, TAIL_CALL_CONNECT4_INDEX);
+
+    // if tail call failed will run this code
+    BPF_LOG(ERR, BACKEND, "workload tail call failed, err is %d\n", ret);
+    return -ENOEXEC;
+}
+
+static inline int backend_manager(ctx_buff_t *ctx, backend_value *backend_v, __u32 service_id, service_value *service_v)
+{
+    int ret;
+    address_t target_addr;
+    __u32 user_port = ctx->user_port;
+
     if (backend_v->waypoint_addr != 0 && backend_v->waypoint_port != 0) {
         BPF_LOG(
             DEBUG,
             BACKEND,
             "find waypoint addr=[%pI4h:%u]\n",
             &backend_v->waypoint_addr,
             bpf_ntohs(backend_v->waypoint_port));
-        value_tuple.ipv4.daddr = ctx->user_ip4;
-        value_tuple.ipv4.dport = ctx->user_port;
-
-        ret = bpf_map_update_elem(&map_of_dst_info, &sk, &value_tuple, BPF_NOEXIST);
-        if (ret) {
-            BPF_LOG(ERR, BACKEND, "record metadata origin address and port failed, ret is %d\n", ret);
+        ret = waypoint_manager(ctx, backend_v->waypoint_addr, backend_v->waypoint_port);
+        if (ret == -ENOEXEC) {
+            BPF_LOG(ERR, BACKEND, "waypoint_manager failed, ret:%d\n", ret);
             return ret;
         }
-        target_addr.ipv4 = backend_v->waypoint_addr;
-        target_addr.port = backend_v->waypoint_port;
-        SET_CTX_ADDRESS(ctx, target_addr);
-        kmesh_workload_tail_call(ctx, TAIL_CALL_CONNECT4_INDEX);
-
-        // if tail call failed will run this code
-        BPF_LOG(ERR, BACKEND, "workload tail call failed, err is %d\n", ret);
-        return -ENOEXEC;
     }
 
 #pragma unroll
-    for (unsigned int i = 0; i < backend_v->port_count; i++) {
-        if (i >= MAX_PORT_COUNT) {
-            BPF_LOG(WARN, BACKEND, "exceed the max port count\n");
+    for (__u32 i = 0; i < backend_v->service_count; i++) {
+        if (i >= MAX_SERVICE_COUNT) {
+            BPF_LOG(WARN, BACKEND, "exceed the max port count:%d\n", MAX_SERVICE_COUNT);
             return -EINVAL;
         }
-
-        if (ctx->user_port == backend_v->service_port[i]) {
-            target_addr.ipv4 = backend_v->ipv4;
-            target_addr.port = backend_v->target_port[i];
-            SET_CTX_ADDRESS(ctx, target_addr);
-            BPF_LOG(
-                DEBUG, BACKEND, "get the backend addr=[%pI4h:%u]\n", &target_addr.ipv4, bpf_ntohs(target_addr.port));
-            return 0;
+        if (service_id == backend_v->service[i]) {
+            BPF_LOG(DEBUG, BACKEND, "access the backend by service:%d\n", service_id);
```

Comment:
rm \n

Reply:
It seems that all logs have been added with ‘\n’. Do we need to delete them all together in this PR.